From 0c524be63ed7c7bf5233db2b6bfad2c62463ab31 Mon Sep 17 00:00:00 2001
From: Tao Guo <tao.guo@amlogic.com>
Date: Wed, 22 May 2019 17:08:23 +0800
Subject: [PATCH] v4l2: fix gstreamer play error [1/1]

PD#SWPL-7623

Problem:
1.v4l2 driver declare it is NV12 format, but output is NV21 format
2.gstreamer skip set format due to v4l2 not support no-contiguous

Solution:
1.Add NV12 and NV21 support. Decoder will set the register depends on the confg
2.Add no-contiguous support

Verify:
U212

Change-Id: Ic2c20c6a2da88da5f5d86d9325cd122af2098f60
Signed-off-by: Tao Guo <tao.guo@amlogic.com>
---
 drivers/amvdec_ports/aml_vcodec_dec.c              | 15 +++++++++++++++
 drivers/frame_provider/decoder/h264_multi/vmh264.c |  8 ++++++--
 2 files changed, 21 insertions(+), 2 deletions(-)

diff --git a/drivers/amvdec_ports/aml_vcodec_dec.c b/drivers/amvdec_ports/aml_vcodec_dec.c
index a2a1b98..4cd1cea 100644
--- a/drivers/amvdec_ports/aml_vcodec_dec.c
+++ b/drivers/amvdec_ports/aml_vcodec_dec.c
@@ -72,10 +72,25 @@ static struct aml_video_fmt aml_video_formats[] = {
 		.num_planes = 1,
 	},
 	{
+		.fourcc = V4L2_PIX_FMT_NV21,
+		.type = AML_FMT_FRAME,
+		.num_planes = 2,
+	},
+	{
 		.fourcc = V4L2_PIX_FMT_NV12,
 		.type = AML_FMT_FRAME,
 		.num_planes = 2,
 	},
+	{
+		.fourcc = V4L2_PIX_FMT_NV21M,
+		.type = AML_FMT_FRAME,
+		.num_planes = 2,
+	},
+	{
+		.fourcc = V4L2_PIX_FMT_NV12M,
+		.type = AML_FMT_FRAME,
+		.num_planes = 2,
+	},
 };
 
 static const struct aml_codec_framesizes aml_vdec_framesizes[] = {
diff --git a/drivers/frame_provider/decoder/h264_multi/vmh264.c b/drivers/frame_provider/decoder/h264_multi/vmh264.c
index 921d46f..b28ccd0 100644
--- a/drivers/frame_provider/decoder/h264_multi/vmh264.c
+++ b/drivers/frame_provider/decoder/h264_multi/vmh264.c
@@ -6519,8 +6519,12 @@ static int vh264_hw_ctx_restore(struct vdec_h264_hw_s *hw)
 #endif
 
 	/* cbcr_merge_swap_en */
-	if (hw->is_used_v4l && v4l2_ctx->ada_ctx->vfm_path
-		!= FRAME_BASE_PATH_V4L_VIDEO)
+	if (hw->is_used_v4l
+		&& v4l2_ctx->ada_ctx->vfm_path != FRAME_BASE_PATH_V4L_VIDEO
+		&& (v4l2_ctx->q_data[AML_Q_DATA_DST].fmt->fourcc
+			== V4L2_PIX_FMT_NV21
+			|| v4l2_ctx->q_data[AML_Q_DATA_DST].fmt->fourcc
+				== V4L2_PIX_FMT_NV21M))
 		SET_VREG_MASK(MDEC_PIC_DC_CTRL, 1 << 16);
 
 	SET_VREG_MASK(MDEC_PIC_DC_CTRL, 0xbf << 24);
-- 
2.7.4

